Mathlib/Algebra/Group/UniqueProds.lean

```diff
@@ -221,149 +241,141 @@ class UniqueProds (G) [Mul G] : Prop where
 
+open UniqueMul in
 @[to_additive] instance {ι} (G : ι → Type*) [∀ i, Mul (G i)] [∀ i, UniqueProds (G i)] :
     UniqueProds (∀ i, G i) where
   uniqueMul_of_nonempty {A} := by
     classical
-    let _ := Finset.isWellFounded_ssubset (α := ∀ i, G i) -- why need this?
+    let _ := isWellFounded_ssubset (α := ∀ i, G i) -- why need this?
     apply IsWellFounded.induction (· ⊂ ·) A; intro A ihA B hA
     apply IsWellFounded.induction (· ⊂ ·) B; intro B ihB hB
-    obtain hc | ⟨i, hc⟩ : (A.card ≤ 1 ∧ B.card ≤ 1) ∨
-      ∃ i, (∃ a1 ∈ A, ∃ a2 ∈ A, a1 i ≠ a2 i) ∨ (∃ b1 ∈ B, ∃ b2 ∈ B, b1 i ≠ b2 i)
-    · obtain hA1 | h1A := le_or_lt A.card 1
-      · obtain hB1 | h1B := le_or_lt B.card 1
-        · exact Or.inl ⟨hA1, hB1⟩
-        obtain ⟨b1, h1, b2, h2, hne⟩ := Finset.one_lt_card.mp h1B
-        obtain ⟨i, hne⟩ := Function.ne_iff.mp hne
-        exact Or.inr ⟨i, Or.inr ⟨b1, h1, b2, h2, hne⟩⟩
-      obtain ⟨a1, h1, a2, h2, hne⟩ := Finset.one_lt_card.mp h1A
-      obtain ⟨i, hne⟩ := Function.ne_iff.mp hne
-      exact Or.inr ⟨i, Or.inl ⟨a1, h1, a2, h2, hne⟩⟩
-    · obtain ⟨a0, ha0⟩ := hA; obtain ⟨b0, hb0⟩ := hB
-      simp_rw [Finset.card_le_one_iff] at hc
-      exact ⟨a0, ha0, b0, hb0, fun a b ha hb _ => ⟨hc.1 ha ha0, hc.2 hb hb0⟩⟩
+    by_cases hc : A.card ≤ 1 ∧ B.card ≤ 1
+    · exact of_card_le_one hA hB hc.1 hc.2
+    simp_rw [not_and_or, not_le] at hc
+    obtain ⟨i, hc⟩ := exists_or.mpr (hc.imp exists_of_one_lt_card_pi exists_of_one_lt_card_pi)
     obtain ⟨ai, hA, bi, hB, hi⟩ := uniqueMul_of_nonempty (hA.image (· i)) (hB.image (· i))
     rw [mem_image, ← filter_nonempty_iff] at hA hB
     let A' := A.filter (· i = ai); let B' := B.filter (· i = bi)
     obtain ⟨a0, ha0, b0, hb0, hu⟩ : ∃ a0 ∈ A', ∃ b0 ∈ B', UniqueMul A' B' a0 b0
-    · rcases hc with ⟨a1, h1, a2, h2, hne⟩ | ⟨b1, h1, b2, h2, hne⟩
-      · refine ihA _ ⟨A.filter_subset _, fun h => ?_⟩ hA hB
-        obtain rfl | hai := eq_or_ne (a1 i) ai
-        exacts [hne (mem_filter.mp <| h h2).2.symm, hai (mem_filter.mp <| h h1).2]
-      by_cases hA' : A' = A
-      · rw [hA']; refine ihB _ ⟨B.filter_subset _, fun h => ?_⟩ hB
-        obtain rfl | hbi := eq_or_ne (b1 i) bi
-        exacts [hne (mem_filter.mp <| h h2).2.symm, hbi (mem_filter.mp <| h h1).2]
-      exact ihA A' ((A.filter_subset _).ssubset_of_ne hA') hA hB
+    · rcases hc with hc | hc; · exact ihA A' (hc.2 ai) hA hB
+      by_cases hA' : A' = A; rw [hA']
+      exacts [ihB B' (hc.2 bi) hB, ihA A' ((A.filter_subset _).ssubset_of_ne hA') hA hB]
     rw [mem_filter] at ha0 hb0
-    refine ⟨a0, ha0.1, b0, hb0.1, fun a b ha hb he => ?_⟩
-    specialize hi (mem_image_of_mem _ ha) (mem_image_of_mem _ hb) ?_
-    · refine (congr_arg (· i) he).trans ?_; rw [← ha0.2, ← hb0.2]; rfl
-    exact hu (mem_filter.mpr ⟨ha, hi.1⟩) (mem_filter.mpr ⟨hb, hi.2⟩) he
+    exact ⟨a0, ha0.1, b0, hb0.1, of_image_filter (Pi.evalMulHom G i) ha0.2 hb0.2 hi hu⟩
+
+open ULift in
+instance [UniqueProds G] [UniqueProds H] : UniqueProds (G × H) := by
```

Comment:
```suggestion
@[to_additive]
instance [UniqueProds G] [UniqueProds H] : UniqueProds (G × H) := by
```
Is there a reason for not additivizing this?  Have I missed it elsewhere? :)

Reply:
Thanks for spotting this! Turns out I need a [shortcut instance](https://github.com/leanprover-community/mathlib4/pull/7169/commits/5e10594daffcae239d69b2a3fef847a2ddc21277#diff-ebdf20b185bac5adc5a22ffdf53e7dfa74bb23fb34dfa47a761bb9dfa66c9fc1R313) to make it work.